plugins/cr8-copy-s3/src/main/java/io/crate/copy/s3/S3ClientHelper.java

```diff
@@ -75,49 +80,93 @@ public AWSCredentials getCredentials() {
+    }
+
+    public AmazonS3 client(URI uri, String protocolSetting) throws IOException {
+        if (protocolSetting == null) {
+            protocolSetting = DEFAULT_PROTOCOL.toString();
+        }
         String accessKey = null;
         String secretKey = null;
-        if (uri.getHost() == null) {
-            throw new IllegalArgumentException(INVALID_URI_MSG);
+        String userInfo = null;
+        if (uri.getHost() == null && uri.getPort() == -1 && uri.getUserInfo() == null) {
+            // userInfo is provided by the user and is contained in authority but NOT in userInfo
+            // this happens when host is not provided
+            if (uri.getAuthority() != null) {
+                int idx = uri.getAuthority().indexOf('@');
+                if (idx != -1) {
+                    userInfo = uri.getAuthority().substring(0, idx);
+                }
+            }
+        } else {
+            userInfo = uri.getUserInfo();
         }
-        if (uri.getUserInfo() != null) {
-            String[] userInfoParts = uri.getUserInfo().split(":");
+        if (userInfo != null) {
+            String[] userInfoParts = userInfo.split(":");
             try {
                 accessKey = userInfoParts[0];
                 secretKey = userInfoParts[1];
             } catch (ArrayIndexOutOfBoundsException e) {
                 // ignore
             }
-            // if the URI contains '@' and ':', a UserInfo is in fact given, but could not
-            // be parsed properly because the URI is not valid (e.g. not properly encoded).
+        // if the URI contains '@' and ':', a UserInfo is in fact given, but could not
+        // be parsed properly because the URI is not valid (e.g. not properly encoded).
         } else if (uri.toString().contains("@") && uri.toString().contains(":")) {
             throw new IllegalArgumentException(INVALID_URI_MSG);
         }
-        return client(accessKey, secretKey);
+
+        String endPoint = null;
+        if (uri.getHost() != null) {
+            endPoint = uri.getHost() + ":" + uri.getPort();
+        }
+
+        return client(accessKey, secretKey, endPoint, protocolSetting);
     }
 
-    private AmazonS3 client(@Nullable String accessKey, @Nullable String secretKey) throws IOException {
-        int hash = hash(accessKey, secretKey);
+    private AmazonS3 client(@Nullable String accessKey, @Nullable String secretKey, @Nullable String endPoint,
+                            String protocolSetting) throws IOException {
+        assert protocolSetting != null : "protocol setting should not be null";
+        int hash = hash(accessKey, secretKey, endPoint, protocolSetting);
         AmazonS3 client = clientMap.get(hash);
         if (client == null) {
-            client = initClient(accessKey, secretKey);
+            client = initClient(accessKey, secretKey, endPoint, protocolSetting);
             clientMap.put(hash, client);
         }
```

Comment:
I think we should change this to have a `ClientKey` class to use as key for the map.

(ClientKey being a record with accessKey, secretKey, endPoint and protocolSetting)

Using the hash of those as key is not working in case of hash collisions.